### ISSUE TITLE
Local (stack) array with 1e7 elements segfaults on macOS

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#5427]
```fortran
integer, parameter :: ten8=1e6
real :: tab(ten8)
call random_number(tab)
print *, 'sum= ', sum(tab), '  expected= ',0.5*ten8
end
```
The above works:
```console
$ lfortran a.f90
sum=     5.00064125e+05      expected=     5.00000000e+05
```
but when you change `ten8` to 1e7, we get a segfault (on macOS):
```console
$ lfortran a.f90
$ echo $?
11
```

**Comment by tanay-man** [comment]
Even with the above example I am getting incorrect results:
```
(lf) tanay@tanay-legion:~/lfortran$ lfortran examples/expr2.f90 
sum=     4.99472594e+05      expected=     5.00000000e+05
```

**Comment by tanay-man** [comment]
This is an equivalent code in C:
```C
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define TEN8 10000000

int main() {
    float tab[TEN8];
    double sum = 0.0;

    srand((unsigned int)time(NULL));

    for (int i = 0; i < TEN8; i++) {
        tab[i] = (float)rand() / RAND_MAX;  // Normalize rand() to [0, 1]
        sum += tab[i];
    }

    // Print the sum and the expected value
    printf("sum = %.2f  expected = %.2f\n", sum, 0.5 * TEN8);

    return 0;
}
```
I am getting a SEGFAULT similar to Lfortran.

**Comment by certik** [comment]
Yes, in C you are explicitly asking for a stack array, so I think it's your responsibility to ensure it can fit.

In Fortran I think we should use stack arrays only if they can fit, otherwise we should use heap arrays. By default I think we should use heap arrays, and only use stack arrays if requested via a compiler option, possibly by default with `--fast`, but the user is then responsible to ensure the arrays fit.

**Comment by tanay-man** [comment]
Ah got it 👍

**Comment by kmr-srbh** [comment]
> In Fortran I think we should use stack arrays only if they can fit, otherwise we should use heap arrays. By default I think we should use heap arrays, and only use stack arrays if requested via a compiler option, possibly by default with --fast, but the user is then responsible to ensure the arrays fit.

Yes, we should be using the heap for this case. GFortran handles it similarly - if the array size is larger than the max var size allowed on stack (defaults to 64kB unless the user increases it at compilation), the arrays are declared on heap. If the value provided by the user exceeds the stack size, they again use the heap.

Reference: https://gcc.gnu.org/onlinedocs/gfortran/Code-Gen-Options.html

This was a very helpful example actually, it revealed a few more things that need to be looked into. I am opening issues for them.

**Comment by HarshitaKalani** [comment]
Works now. Let's add a test to close this.
With 1e6
```console
% gfortran a.f90 && ./a.out
 sum=    500597.625       expected=    500000.000    
% lfortran a.f90 
sum=     500262.469      expected=     500000.000
```
! With 1e7
```console
% lfortran a.f90
sum=     4999023.00      expected=     5000000.00
% gfortran a.f90 && ./a.out
 sum=    4998559.50       expected=    5000000.00  
```
